server/s3_bucket_metadata_cache.cc

```diff
@@ -0,0 +1,364 @@
+
+  s3_log(S3_LOG_DEBUG, nullptr, "%s Exit", __func__);
+}
+
+void S3BucketMetadataCache::Item::update(const S3BucketMetadata& src,
+                                         StateHandlerType callback) {
+  s3_log(S3_LOG_DEBUG, nullptr, "%s Entry", __func__);
+
+  if (current_op != CurrentOp::none) {
+    s3_log(S3_LOG_ERROR, "", "Another operation is beeing made");
+    callback(S3BucketMetadataState::failed);
+    return;
+  }
+  current_op = CurrentOp::saving;
+  on_changed = std::move(callback);
+
+  create_engine(src);
+  p_engine->update(src, std::bind(&S3BucketMetadataCache::Item::on_done, this,
+                                  std::placeholders::_1));
+
+  s3_log(S3_LOG_DEBUG, nullptr, "%s Exit", __func__);
+}
+
+void S3BucketMetadataCache::Item::remove(StateHandlerType callback) {
+  s3_log(S3_LOG_DEBUG, nullptr, "%s Entry", __func__);
+
+  if (current_op != CurrentOp::none) {
+    s3_log(S3_LOG_ERROR, "", "Another operation is beeing made");
+    callback(S3BucketMetadataState::failed);
+    return;
+  }
+  current_op = CurrentOp::deleting;
+  on_changed = std::move(callback);
+
+  create_engine(*p_value);
+  p_engine->remove(std::bind(&S3BucketMetadataCache::Item::on_done, this,
+                             std::placeholders::_1));
+
+  s3_log(S3_LOG_DEBUG, nullptr, "%s Exit", __func__);
+}
+
+// ************************************************************************* //
+// class S3BucketMetadataCache
+// ************************************************************************* //
+
+S3BucketMetadataCache::S3BucketMetadataCache(
+    unsigned max_cache_size, unsigned expire_interval_sec,
+    unsigned refresh_interval_sec,
+    std::shared_ptr<S3MotrBucketMetadataFactory> motr_bucket_metadata_factory)
+    : max_cache_size(max_cache_size),
+      expire_interval_sec(expire_interval_sec),
+      refresh_interval_sec(refresh_interval_sec) {
+
+  if (p_bucket_metadata_cache) {
+    s3_log(S3_LOG_FATAL, "",
+           "Only one instance of S3BucketMetadataCache is allowed");
+  }
+  this->s3_motr_bucket_metadata_factory =
+      motr_bucket_metadata_factory
+          ? std::move(motr_bucket_metadata_factory)
+          : std::make_shared<S3MotrBucketMetadataFactory>();
+  p_bucket_metadata_cache = this;
+}
+
+S3BucketMetadataCache::~S3BucketMetadataCache() {
+  p_bucket_metadata_cache = nullptr;
+}
+
+void S3BucketMetadataCache::remove_item(std::string bucket_name) {
+
+  assert(sorted_by_access.size() == sorted_by_update.size());
+  assert(sorted_by_access.size() == items.size());
+
+  auto map_it = items.find(bucket_name);
+  assert(items.end() != map_it);
+
+  auto* p_item = (*map_it).second.get();
+
+  sorted_by_access.erase(p_item->ptr_access);
+  sorted_by_update.erase(p_item->ptr_update);
+  items.erase(map_it);
+
+  s3_log(S3_LOG_DEBUG, "",
+         "Metadata for \"%s\" has been removed from the cache",
+         bucket_name.c_str());
+}
+
+void S3BucketMetadataCache::shrink() {
```

Comment:
Here is preliminary, simplified logic now.

Reply:
Discussed with Evgeniy and decided:

1. Need to add similar code to first try and find expired entry which can be removed.
2. Need to add comment in Part 2 code which can allow us to "block" a request which caused "pre-load" logic, so that it waits for the response.  This way we can guarantee that count of motr requests is not higher than count of S3 API requests.
3. Corner case: code requests bucket metadata, and all entries in cache are locked.  In this case -- add warning to log (as already done below), AND return error "Come back later" to API client.

All this should be done in Part 2, so closing this note.